macaroon_service.go

```diff
@@ -231,12 +234,46 @@ func (ms *MacaroonService) Start() error {
 
 	// There are situations in which we don't want a macaroon to be created
 	// on disk (for example when running inside LiT stateless integrated
-	// mode). For any other cases, we create macaroon files in the default
-	// directory.
-	if ms.cfg.StatelessInit || lnrpc.FileExists(ms.cfg.MacaroonPath) {
+	// mode).
+	if ms.cfg.StatelessInit {
 		return nil
 	}
 
+	// If there are no required permissions, then we don't have to generate
+	// or re-generate a macaroon and so we can exit here.
+	if len(ms.cfg.RequiredPerms) == 0 {
+		return nil
+	}
+
+	// If we are not in stateless mode and a macaroon file does exist, we
+	// check that the macaroon matches the required permissions. If not, we
+	// will delete the macaroon and create a new one.
+	if lnrpc.FileExists(ms.cfg.MacaroonPath) {
+		matches, err := ms.macaroonMatchesPermissions()
+		if err != nil {
+			return fmt.Errorf("unable to match the previous "+
+				"macaroon's permissions with the required "+
+				"permissions: %v", err)
+		}
+
+		// In case the old macaroon matches the required permissions,
+		// we don't need to create a new macaroon.
+		if matches {
+			return nil
+		}
+
+		// Else if the required permissions have been updated, we delete
+		// the old macaroon and create a new one.
+		log.Infof("Macaroon at %s does not have all required "+
+			"permissions. Deleting it and creating a new "+
+			"one", ms.cfg.MacaroonPath)
+
+		if err := os.Remove(ms.cfg.MacaroonPath); err != nil {
+			return fmt.Errorf("Unable to remove macaroon "+
+				"at %s: %v", ms.cfg.MacaroonPath, err)
+		}
+	}
+
```

Comment:
also , while we are here, do you mind replacing the `ioutil.WriteFile` below with `os.WriteFile`?